homeassistant/components/sma/__init__.py

```diff
@@ -1 +1,199 @@
-"""The sma component."""
+"""The sma integration."""
+import asyncio
+from datetime import timedelta
+import logging
+
+import pysma
+
+from homeassistant.config_entries import SOURCE_IMPORT, ConfigEntry, ConfigEntryNotReady
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_PATH,
+    CONF_SCAN_INTERVAL,
+    CONF_SENSORS,
+    CONF_SSL,
+    CONF_VERIFY_SSL,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import (
+    CONF_CUSTOM,
+    CONF_FACTOR,
+    CONF_GROUP,
+    CONF_KEY,
+    CONF_UNIT,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+    PLATFORMS,
+    PYSMA_COORDINATOR,
+    PYSMA_OBJECT,
+    PYSMA_REMOVE_LISTENER,
+    PYSMA_SENSORS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _parse_legacy_options(entry: ConfigEntry, sensor_def: pysma.Sensors) -> None:
+    """Parse legacy configuration options.
+
+    This will parse the legacy CONF_SENSORS and CONF_CUSTOM configuration options
+    to support deprecated yaml config from platform setup.
+    """
+
+    # Add sensors from the custom config
+    sensor_def.add(
+        [
+            pysma.Sensor(o[CONF_KEY], n, o[CONF_UNIT], o[CONF_FACTOR], o.get(CONF_PATH))
+            for n, o in entry.data.get(CONF_CUSTOM).items()
+        ]
+    )
+
+    # Parsing of sensors configuration
+    config_sensors = entry.data.get(CONF_SENSORS)
+    if not config_sensors:
+        return
+
+    # Find and replace sensors removed from pysma
+    # This only alters the config, the actual sensor migration takes place in _migrate_old_unique_ids
+    for sensor in config_sensors.copy():
+        if sensor in pysma.LEGACY_MAP:
+            config_sensors.remove(sensor)
+            config_sensors.append(pysma.LEGACY_MAP[sensor]["new_sensor"])
+
+    # Only sensors from config should be enabled
+    for sensor in sensor_def:
+        sensor.enabled = sensor.name in config_sensors
+
+
+async def _migrate_old_unique_ids(
+    hass: HomeAssistant, entry: ConfigEntry, sensor_def: pysma.Sensors
+) -> None:
+    """Migrate legacy sensor entity_id format to new format."""
+    entity_registry = er.async_get(hass)
+
+    # Create list of all possible sensor names
+    possible_sensors = list(
```

Comment:
We don't need to copy to list it seems. We can iterate the set.

Reply:
Yes you are right. I have tested this to confirm it is working by iterating the set.